script/modelling/maxcut.jl

```diff
@@ -0,0 +1,169 @@
+@show objective_value(linear_max_cut);
+
+x_linear = value.(x)
+@show x_linear;
+@show [(i,j) for i in 1:n-1 for j in i+1:n if JuMP.value.(z)[i,j] > 0.5];
+
+#' Visualizing a solution
+
+nodecolor = [colorant"lightseagreen", colorant"orange"]
+all_node_colors = [nodecolor[round(Int, xi + 1)]  for xi in x_linear]
+GraphPlot.gplot(g, nodelabel=1:6, nodefillc=all_node_colors)
+
+#' ### Semidefinite formulation
+
+#' The maximum cut problem can also be formulated as a quadratic optimization model.
+#' If the solution vector is composed of $\{-1, 1\}$ for belonging to the subset or not.
+#' Then, the product $x_i x_j$ is $-1$ if i and j are in different subsets, and 1 otherwise.
+#' This implies $1 - x_i x_j$ is $0$ if i and j are in the same subset, and 2 otherwise.
+#' Since each pair is counted twice with (i,j) and (j,i), we need to divide it by 4:
+#' $\frac{w_{ij}}{4} (1 - x_i x_j)$ is $0$ if i and j are in the same subset, and $z_{ij}$ otherwise.
+
+#' The resulting model is given by:
+#' $$
+#' \begin{align*}
+#' \max_{x} & \frac{w_{ij}}{4} (1 - x_i x_j) \\
+#' \text{s.t. } & x_i \in \{-1, 1\} \forall i \in V
+#' \end{align*}
+#' $$
+
+#' Introducing a matrix variable $Y_{ij} = x_i x_j$,
+#' we can equivalently reformulate the above quatratic problem as follows:
+#' $$
+#' \begin{align*}
+#' \max_{x,Y} & \frac{w_{ij}}{4} (1 - Y_{ij}) \\
+#' \text{s.t. } & Y_{ii} = 1 \forall i \in V \\
+#'              & Y = x x^T
+#' \end{align*}
+#' $$
+
+#' The constraint $Y = x x^T$ requires $Y$ to be a Positive Semidefinite matrix
+#' of rank one. This optimization problem is fully equivalent to the initial
+#' quadratic problem and hard to solve.
+#' A possible relaxation is to remove the rank-1 constraint, yielding the following problem:
+#' $$
+#' \begin{align*}
+#' \max_{Y} & \frac{w_{ij}}{4} (1 - Y_{ij}) \\
+#' \text{s.t. } & Y_{ii} = 1 \forall i \in V \\
+#'              & Y \in \mathcal{S}_n^+
+#' \end{align*}
+#' $$
+
+#' with $\mathcal{S}_n^+$ the set of positive semidefinite matrices.
+#' Such model can be implemented in JuMP and solved using an underlying solver
+#' which supports semidefinite optimization.
+
+sdp_max_cut = Model(optimizer_with_attributes(SCS.Optimizer, "verbose" => 0))
+
+@variable(sdp_max_cut, Y[1:n,1:n] in PSDCone())
+@constraint(sdp_max_cut, [i = 1:n], Y[i,i] == 1)
+@objective(sdp_max_cut, Max, 1/4 * sum(w[i,j] * (1 - Y[i,j]) for i in 1:n for j in 1:n))
+
+optimize!(sdp_max_cut)
+
+@test objective_value(linear_max_cut) >= 6.0
+@show objective_value(linear_max_cut);
+
+#' Re-computing the vector:
+#' $Y$ approximates $x x^T$, we can compute the estimate by randomized rounding.
+#' Goemans, M. X., & Williamson, D. P. (1995).
+#' Improved approximation algorithms for maximum cut and satisfiability problems using semidefinite programming.
+#' Journal of the ACM (JACM), 42(6), 1115-1145.
+
+F = svd(value.(Y))
+U = F.U * Diagonal(sqrt.(F.S))
+
+Random.seed!(33)
+x = rand(size(U, 2)) .- 0.5 # random vector of mean 0
```

Comment:
The issue with this is that it generates a point uniformly in the hypercube so it's not an uniformly sampled direction.
As you don't care about the norm of x, you can assume that it's norm is normally distributed. As exp(x1^2 + x2^2) = exp(x1^2) exp(x2^2), you see that you can sample each coordinate normally and it will sample the norm normally!
So the solution is to sample each coordinate with a univariate standard uniform distribution

Reply:
thanks indeed I thought only having centered sample was enough, updated in latest commit